staging/src/k8s.io/apiserver/plugin/pkg/kmsprovider/vault/client_test.go

```diff
@@ -0,0 +1,419 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package vault implements envelop encryption provider based on Vault KMS
+package vault
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/vault/api"
+	"github.com/pborman/uuid"
+)
+
+const (
+	cafile     = "testdata/ca.crt"
+	serverCert = "testdata/server.crt"
+	serverKey  = "testdata/server.key"
+	clientCert = "testdata/client.crt"
+	clientKey  = "testdata/client.key"
+)
+
+func TestTypicalCase(t *testing.T) {
+	server := VaultTestServer(t, nil)
+	defer server.Close()
+
+	configToken := &EnvelopeConfig{
+		Token:   uuid.NewRandom().String(),
+		Address: server.URL,
+		CACert:  cafile,
+	}
+	configTLS := &EnvelopeConfig{
+		ClientCert: clientCert,
+		ClientKey:  clientKey,
+		Address:    server.URL,
+		CACert:     cafile,
+	}
+	configRole := &EnvelopeConfig{
+		RoleID:  uuid.NewRandom().String(),
+		Address: server.URL,
+		CACert:  cafile,
+	}
+
+	configs := []struct {
+		name   string
+		config *EnvelopeConfig
+	}{
+		{"AuthByToken", configToken},
+		{"AuthByTLS", configTLS},
+		{"AuthByAppRole", configRole},
+	}
+
+	for _, testCase := range configs {
+		encryptAndDecrypt(t, testCase.name, testCase.config)
+	}
+}
+
+func TestCustomTransitPath(t *testing.T) {
+	customPath := "custom-transit"
+	server := customTransitPathServer(t, customPath)
+	defer server.Close()
+
+	config := &EnvelopeConfig{
+		Token:   uuid.NewRandom().String(),
+		Address: server.URL,
+		CACert:  cafile,
+	}
+
+	// Work with custom transit path
+	validPathes := []string{customPath, "/" + customPath, customPath + "/", "/" + customPath + "/"}
+	for _, path := range validPathes {
+		config.TransitPath = path
+		encryptAndDecrypt(t, path, config)
+	}
+
+	// Invalid transit path will result 404 error
+	config.TransitPath = "invalid-" + customPath
+	client, err := newClientWrapper(config)
+	if err != nil {
+		t.Fatalf("fail to initialize Vault client: %s ", err)
+	}
+
+	_, err = client.encrypt("key", "plain")
+	if err == nil || !strings.Contains(err.Error(), "404") {
```

Comment:
anyway to avoid `strings.Contains` here?

Reply:
The vault API used does not have any specific error interface. We have to check the error message to assert the behavior. The message is expected to be small and this is testcase code, if your concern is the performance.